### PR TITLE
ASoC: SOF: add file names and paths for IPCv4 support

### PR DESCRIPTION
**Pull request by plbossart**
This PR does not change any functionality for existing platforms, it only prepares the support of IPCv4 which will require a different userspace setup (firmware path, firmware name, topology path).

The layout for IPCv4 will be the following (using tgl as a platform example)

/lib/intel/avs/tgl
   dsp_basefw.bin

### CODE REVIEW

File: sound/soc/sof/sof-pci-dev.c [hunk 1]
```diff
@@ -178,22 +178,22 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
...
-			sof_pdata->desc->default_fw_path;
+			sof_pdata->desc->default_fw_path[SOF_IPC];
 	}
 
 	if (tplg_path)
 		sof_pdata->tplg_filename_prefix = tplg_path;
 	else
 		sof_pdata->tplg_filename_prefix =
-			sof_pdata->desc->default_tplg_path;
+			sof_pdata->desc->default_tplg_path[SOF_IPC];
```
Comment (ranj063) [inline sound/soc/sof/sof-pci-dev.c]: did you miss sof-of-dev.c @plbossart?

File: include/sound/sof.h [hunk 2]
```diff
@@ -122,6 +122,10 @@ struct sof_dev_desc {
 	/* defaults for no codec mode */
 	const char *nocodec_tplg_filename;
 
+	/* information on supported IPCs */
```
Comment (ranj063) [inline include/sound/sof.h:127]: you mean SOF_INTEL_IPC4 instead of cavs IPC in the commit message?

File: sound/soc/sof/intel/pci-apl.c [hunk 3]
```diff
@@ -37,7 +37,10 @@ static const struct sof_dev_desc bxt_desc = {
 		[SOF_IPC] = "intel/sof-tplg",
 		[SOF_INTEL_IPC4] = "intel/avs-tplg",
 	},
-	.default_fw_filename = "sof-apl.ri",
+	.default_fw_filename = {
+		[SOF_IPC] = "sof-apl.ri",
+		[SOF_INTEL_CAVS] = "dsp_basefw.bin",
```
Comment (ranj063) [inline sound/soc/sof/intel/pci-apl.c]: SOF_INTEL_IPC4?

File: sound/soc/sof/sof-pci-dev.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: fixed

File: include/sound/sof.h (hunk 2 above)
Comment (plbossart) [reply include/sound/sof.h:127]: yes, nice catch. fixed now.

File: sound/soc/sof/intel/pci-apl.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/pci-apl.c]: fixed

File: sound/soc/sof/sof-pci-dev.c [hunk 4]
```diff
@@ -158,7 +158,9 @@ int sof_pci_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
 	sof_pdata->name = pci_name(pci);
 	sof_pdata->desc = desc;
 	sof_pdata->dev = dev;
-	sof_pdata->fw_filename = desc->default_fw_filename;
+
+	sof_pdata->ipc_type = desc->ipc_default;
```
Comment (bardliao) [inline sound/soc/sof/sof-pci-dev.c:162]: Should `sof_pdata->ipc_type = desc->ipc_default;` belong to `ASoC: SOF: add default IPC capability and file paths`?

File: include/sound/sof.h [hunk 5]
```diff
@@ -124,6 +124,10 @@ struct sof_dev_desc {
...
+	int ipc_supported_mask;
+	int ipc_default;
```
Comment (RanderWang) [inline include/sound/sof.h]: ipc_default is a enum so uint is better. And also ipc_supported_mask

Comment (RanderWang) [reply include/sound/sof.h]: or define it as "enum sof_ipc_type ipc_default" so user can get more info easy.

File: sound/soc/sof/sof-pci-dev.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c:162]: I wanted the "ASoC: SOF: add default IPC capability and file paths" to only change the tables - this makes the review easier. 
The selection of the IPC type based on the default is a change of functionality, so belongs to the next patch I believe.

File: include/sound/sof.h (hunk 5 above)
Comment (plbossart) [reply include/sound/sof.h]: done

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by ujfalusi — APPROVED** [review]
@plbossart, looks neat and clean.

One question: are you planning to add support to override the default IPC version for platforms which can work with different versions of IPC?

**Comment by plbossart** [comment]
> @plbossart, looks neat and clean.
> 
> One question: are you planning to add support to override the default IPC version for platforms which can work with different versions of IPC?

yes, it's done and tested in my local branch. I tested the closed-source firmware on TGL. It's not helpful though until we can override the .ops.

I will also need to add an option to override the file name from the command line.

**Comment by aiChaoSONG** [comment]
@plbossart For JSL with IPC3, the firmware name was sof-jsl.ri, and now it is sof-icl.ri, is this expected change?

**Comment by plbossart** [comment]
> @plbossart For JSL with IPC3, the firmware name was sof-jsl.ri, and now it is sof-icl.ri, is this expected change?

no, this is a copy/paste mistake

````
	.default_fw_filename = "sof-jsl.ri",
	.default_fw_filename = {
		[SOF_IPC] = "sof-icl.ri",
		[SOF_INTEL_IPC4] = "dsp_basefw.bin",
	},
````

will send a fixup tomorrow. Thanks for spotting this @aiChaoSONG, much appreciated.

**Pull request merged**